### PR TITLE
Handle responses with no headers

### PR DESCRIPTION
**Pull request by dragosMC91**
This is an attempt to address the issue mentioned here: https://github.com/webdriverio-community/wdio-intercept-service/issues/416.

## Issue summary
Some requests have empty `response headers` and the new implementation of `parseResponseHeaders()` doesn't handle this case.
For the project i am working on this is the case with requests from [mapbox](https://www.mapbox.com/).

## Issue details
While inspecting these requests in the chrome `Network` tab, the headers are
```
access-control-allow-origin: *
date: Thu, 01 Dec 2022 11:31:03 GMT
x-powered-by: Express
```
However, while trying to complete the fetch request here: https://github.com/webdriverio-community/wdio-intercept-service/blob/main/lib/interceptor.js#L104, `clonedResponse.headers` has no entries, the request after this point looks like this:
```
{
	"method":"POST",
	"requestHeaders":"content-type: text/plain",
	"requestBody":"[{\\"event\\":\\"map.load\\",\\"created\\":\\"2022-12-01T11:31:03.293Z\\",\\"sdkIdentifier\\":\\"mapbox-gl-js\\",\\"sdkVersion\\":\\"2.4.1\\",\\"skuId\\":\\"01\\",\\"userId\\":\\"redacted\\",\\"skuToken\\":\\"redacted\\"}]",
	"url":"https://events.mapbox.com/events/v2?access_token=redacted","__processed":1669894263694,
	"body":"",
	"headers":"",
	"statusCode":204,
	"__fulfilled":1669894263831
}
```
 This means `parseHeaders()` returns an empty string which generates the mentioned error when passed to `parseResponseHeaders()`. 
 
 ## Proposed fix
 Considering the situation where the headers might be empty was already taken into account in the interceptor: https://github.com/webdriverio-community/wdio-intercept-service/blob/main/lib/interceptor.js#L200, this should also be handled by `parseResponseHeaders` here https://github.com/webdriverio-community/wdio-intercept-service/blob/main/index.js#L345.
 
 I tried adding a test for this as well but i didn't manage to create a `XMLHttpRequest` request which responded with empty headers so i used the wdio.mock API which unfortunately only works on chrome.

### CODE REVIEW

**Review by christian-bromann — APPROVED** [review]
LGTM 👍 

@tehhowch thoughts?

File: index.js [hunk 1]
```diff
@@ -344,6 +344,9 @@ class WebdriverAjax {
     // (best effort compliance with RFC)
     function parseResponseHeaders(rawHeader) {
       const headers = {};
+      if (rawHeader === '') {
+        return headers;
```
Comment (tehhowch) [inline index.js:348]: Can we get some logging for this scenario, e.g. `"${transformed.method} request to ${req.url} (HTTP ${req.statusCode}) had no response headers!"`, [here](https://github.com/webdriverio-community/wdio-intercept-service/blob/0162292cb2bb7c0761cefb0f37881e233ff0fb17/index.js#L330-L331)? That way people can at least see how often this happens. We could suggest reporting the request here, too.

**Review by tehhowch — COMMENTED** [review]
I still have some trouble understanding how we get to a point where there are _no response headers at all_ without there being some precursor bug in how we extract headers from the intercepted request object. As you yourself have said,

> I didn't manage to create a XMLHttpRequest request which responded with empty headers

I too expect there to basically always be some response headers, especially for an HTTP 200. Maybe an HTTP 204 (no content) could be conceivably response-header-free?

**Comment by dragosMC91** [comment]
Hello @tehhowch, thank you for the feedback! 

I too thought there is
> some precursor bug in how we extract headers from the intercepted request object

and spent some time investigating this but couldn't find it, i will do a bit more digging now that i'm more familiar with the library and post my findings here. If i still don't find the bug i will implement the above suggestions.

I'll get to my pc this evening and complete the above.

**Comment by dragosMC91** [comment]
@tehhowch i fiddled some more with the interceptor but i didn't find anything, for example when `addPendingRequest()` is called here https://github.com/webdriverio-community/wdio-intercept-service/blob/0162292cb2bb7c0761cefb0f37881e233ff0fb17/lib/interceptor.js#L92 the `headers` are empty and at this point there has been no processing done on the headers from what i understand.

DevTools - www zameen com_Property_dha_defence_dha_defence_phase_2_owner_bu  2022-12-03 at 10 14 44 PM

This means the bug is not coming from the interceptor but maybe im missing something.

Also, addressed the comments about the usage of `assert` instead of `expect` and adding a console warning in the last commit.

File: index.js [hunk 2]
```diff
@@ -344,6 +349,9 @@ class WebdriverAjax {
     // (best effort compliance with RFC)
     function parseResponseHeaders(rawHeader) {
       const headers = {};
+      if (rawHeader === '') {
```
Comment (tehhowch) [inline index.js:352]: `XMLHttpRequest.getAllResponseHeaders()` can return `null` in exceptional cases
```suggestion
      if (!rawHeader) {
```

**Review by tehhowch — APPROVED** [review]
Thanks for adding the logging bit! I took a look at the implementation of `parseResponseHeaders` prior to #241 and confirmed that it would return `null` if the input was null, or an empty object if the input was an empty string, so the guard you're adding here is restoring previous functionality.

Could you update the changelog file with your change?

I also looked at mapbox briefly; they apparently have a mode that supports automated testing and doesn't require sharing your [access token](https://docs.mapbox.com/mapbox-gl-js/guides/browsers-and-testing/#automated-tests). If you wanted to put together a gist that generates a request that fails without your change, that could be useful for additional debugging. (The part that is curious to me is that the request you have shared is HTTP 204, i.e. "No Content"). You could share the gist on the backing issue for this PR, #416

**Comment by dragosMC91** [comment]
@tehhowch i made the requested updates here: https://github.com/webdriverio-community/wdio-intercept-service/pull/438/commits/89d64d8ae1ad79c65d7db346496c39de50909d0f

Also, added a reproductible example for the bug here: https://github.com/webdriverio-community/wdio-intercept-service/issues/416#issuecomment-1336526124

**Comment by dragosMC91** [comment]
hello @tehhowch, is this going to get merged ? saw another person complaining about the bug https://github.com/webdriverio-community/wdio-intercept-service/issues/416#issuecomment-1351056749
and I too would like to update to latest :D

**Review by tehhowch — APPROVED** [review]
Thanks for putting together that example! Seems like HTTP 204's from mapbox are particularly prone to having no response headers.

I'd like to clean up the test to work on firefox, too, but that's something we can achieve separately.
